File: anki/anki.go
package anki

import (
	"bufio"
	"database/sql"
	"fmt"
	"os"
	"regexp"
	"strings"

	catppuccin "github.com/catppuccin/go"
	"github.com/charmbracelet/lipgloss"
)

func ReadLines(path string) ([]string, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		// check if line starts with # or is empty
		if len(scanner.Text()) == 0 || scanner.Text()[0] == '#' {
			continue
		}

		lines = append(lines, scanner.Text())
	}

	return lines, scanner.Err()
}

type AnkiCard struct {
	Word         string
	Reading      string
	Translations []string
}

func RemoveHTMLTags(s string) string {
	regexp := regexp.MustCompile("<[^>]*>")
	return regexp.ReplaceAllString(s, "")
}

func ParseLine(line string) (AnkiCard, error) {
	// split line by x1f
	parts := strings.Split(line, "\x1f")

	// check if line has 3 parts
	if len(parts) < 4 {
		return AnkiCard{}, fmt.Errorf("line has %d parts, at least 4", len(parts))
	}

	translations := strings.Split(parts[3], "<br>")

	// trim & remove quotes
	for i, translation := range translations {
		translations[i] = strings.Trim(translation, "\"")
		translations[i] = strings.TrimSpace(translations[i])
	}

	return AnkiCard{
		Word:         parts[1],
		Reading:      parts[2],
		Translations: translations,
	}, nil
}

func RandomCard(db *sql.DB, deck_id uint64) (AnkiCard, error) {
	rows, err := db.Query("SELECT notes.flds FROM cards JOIN notes ON cards.nid = notes.id WHERE cards.did = ? ORDER BY random() LIMIT 1", deck_id)
	if err != nil {
		return AnkiCard{}, err
	}
	defer rows.Close()

	var flds string
	if rows.Next() {
		err = rows.Scan(&flds)
		if err != nil {
			return AnkiCard{}, err
		}
	}

	return ParseLine(flds)
}

// func RandomCard(path string) (AnkiCard, error) {
// 	lines, err := ReadLines(path)
// 	if err != nil {
// 		return AnkiCard{}, err
// 	}

// 	// pick a random line
// 	idx := rand.New(rand.NewSource(time.Now().UnixNano())).Intn(len(lines))
// 	line := lines[idx]

// 	return ParseLine(line)
// }

func RenderTranslation(translation string) string {
	styleType := lipgloss.NewStyle().Foreground(lipgloss.Color(catppuccin.Mocha.Subtext0().Hex)).Italic(true)
	styleNumber := lipgloss.NewStyle().
		Background(lipgloss.Color(catppuccin.Mocha.Pink().Hex)).
		Foreground(lipgloss.Color(catppuccin.Mocha.Crust().Hex)).
		Width(3).
		Align(1).
		MarginRight(1)
	styleWord := lipgloss.NewStyle()

	if strings.Contains(translation, "78909C") {
		// word type
		translation = RemoveHTMLTags(translation)
		translation = strings.TrimSpace(translation)
		translation = styleType.Render(translation)
	} else {
		// translation
		split := strings.Split(translation, "</b>")

		if len(split) < 2 {
			translation = RemoveHTMLTags(translation)
			translation = strings.TrimSpace(translation)
			translation = styleWord.Render(translation)
		} else {
			// first element is the number of the translation
			number := split[0]
			number = RemoveHTMLTags(number)
			number = strings.TrimSpace(number)

			translation = strings.Join(split[1:], "")
			translation = RemoveHTMLTags(translation)
			translation = strings.TrimSpace(translation)

			translation = styleNumber.Render(number) + styleWord.Render(translation)
		}
	}

	return translation
}

func RenderCard(card AnkiCard) string {
	styleWord := lipgloss.NewStyle().Foreground(lipgloss.Color(catppuccin.Mocha.Pink().Hex))
	styleReading := lipgloss.NewStyle().Foreground(lipgloss.Color(catppuccin.Mocha.Subtext0().Hex))

	translations := ""
	for _, translation := range card.Translations {
		translations += RenderTranslation(translation) + "\n"
	}

	out := lipgloss.JoinVertical(0,
		styleWord.Render(card.Word)+" - "+styleReading.Render(card.Reading),
		translations,
	)

	// border := lipgloss.NewStyle().
	// 	BorderStyle(lipgloss.NormalBorder()).
	// 	BorderForeground(lipgloss.Color(catppuccin.Mocha.Overlay2().Hex)).
	// 	BorderTop(true)

	return out
}

File: src/anki.rs
use itertools::Itertools;
use rand::seq::SliceRandom;
use rusqlite::Connection;
use yansi::{Color, Paint, Style};

pub(crate) fn random_card(
    connection: Connection,
    deck_id: u64,
) -> Result<AnkiCard, Box<dyn std::error::Error>> {
    let mut stmt = connection.prepare(
        "SELECT notes.flds FROM cards
    JOIN notes ON cards.nid = notes.id
    WHERE cards.did = ?",
    )?;

    let rows = stmt
        .query_map(&[&deck_id], |row| row.get(0))?
        .filter(|r| r.is_ok())
        .map(|r| r.unwrap())
        .collect::<Vec<String>>();

    let card = rows.choose(&mut rand::thread_rng()).unwrap();

    Ok(parse_line(card)?)
}

#[derive(Debug)]
pub(crate) struct AnkiCard {
    word: String,
    reading: String,
    translations: Vec<String>,
}

fn parse_line(line: &str) -> Result<AnkiCard, &'static str> {
    let mut parts = line.split('\x1f');

    let err = "malformed anki card";

    dbg!(line);

    parts.next().ok_or(err)?;

    let word = parts.next().ok_or(err)?.to_string();
    let reading = parts.next().ok_or(err)?.to_string();
    let translations = parts
        .next()
        .ok_or(err)?
        .split("<br>")
        .map(|s| s.trim().to_string())
        .collect();

    Ok(AnkiCard {
        word,
        reading,
        translations,
    })
}

trait RemoveHtmlTags<O> {
    fn remove_html_tags(&self) -> O;
}

impl RemoveHtmlTags<String> for &str {
    fn remove_html_tags(&self) -> String {
        let mut in_tag = false;
        let mut result = String::new();

        for c in self.chars() {
            if c == '<' {
                in_tag = true;
            } else if c == '>' {
                in_tag = false;
            } else if !in_tag {
                result.push(c);
            }
        }

        result
    }
}

impl RemoveHtmlTags<String> for String {
    fn remove_html_tags(&self) -> String {
        self.as_str().remove_html_tags()
    }
}

const fn color_ctp_to_yansi(color: catppuccin::Color) -> Color {
    Color::Rgb(color.rgb.r, color.rgb.g, color.rgb.b)
}

static COLOR_CRUST: Color = color_ctp_to_yansi(catppuccin::PALETTE.mocha.colors.crust);
static COLOR_SUBTEXT0: Color = color_ctp_to_yansi(catppuccin::PALETTE.mocha.colors.subtext0);
static COLOR_PINK: Color = color_ctp_to_yansi(catppuccin::PALETTE.mocha.colors.pink);

fn render_translation(translation: &str) -> String {
    let mut translation = translation.to_string();

    static STYLE_TYPE: Style = Style::new().fg(COLOR_SUBTEXT0).italic();
    static STYLE_NUMBER: Style = Style::new().bg(COLOR_PINK).fg(COLOR_CRUST); // TODO size & alignment
    static STYLE_WORD: Style = Style::new();

    if translation.contains("78909C") {
        translation = translation
            .remove_html_tags()
            .trim()
            .paint(STYLE_TYPE)
            .to_string();
    } else {
        let split: Vec<&str> = translation.splitn(2, "</b>").collect();

        if split.len() == 2 {
            let (number, word) = (split[0], split[1]);

            let number = number.remove_html_tags().trim().to_owned();

            let number = format!("{}{}", " ".repeat(3 - number.len()), number)
                .paint(STYLE_NUMBER)
                .to_string();

            let word = word.remove_html_tags().trim().paint(STYLE_WORD).to_string();

            translation = format!("{} {}", number, word);
        } else {
            translation = translation.paint(STYLE_WORD).to_string();
        }
    }

    return translation;
}

impl AnkiCard {
    pub(crate) fn render(&self) -> String {
        static STYLE_WORD: Style = Style::new().fg(COLOR_PINK);
        static STYLE_READING: Style = Style::new().fg(COLOR_SUBTEXT0);

        let translations = self
            .translations
            .iter()
            .map(|t| render_translation(&t))
            .join("\n");

        format!(
            "{} - {}\n{}",
            self.word.paint(STYLE_WORD),
            self.reading.paint(STYLE_READING),
            translations
        )
    }
}
